src/main/scala/mesosphere/util/ZookeeperDirectState.scala

```diff
@@ -0,0 +1,186 @@
+package mesosphere.util
+
+import java.lang.Boolean
+import java.util
+import java.util.UUID
+import java.util.concurrent.{ Future => JFuture }
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, Future, TimeoutException }
+import scala.util.{ Failure, Success }
+
+import com.google.protobuf.ByteString
+import mesosphere.marathon.Protos
+import org.apache.curator.framework._
+import org.apache.curator.retry.ExponentialBackoffRetry
+import org.apache.curator.utils.ZKPaths
+import org.apache.mesos.state.{ State, Variable }
+import org.apache.zookeeper.data.Stat
+
+/**
+  * A type of State that persists data in the form of a mesosphere.marathon.Protos.Entry
+  * protocol buffer message to Zookeeper.
+  *
+  * Does not use JNI or the native libmesos library, unlike "ZooKeeperState"
+  */
+class ZookeeperDirectState(
+  zNode: String,
+  serverList: String,
+  sessionTimeout: Duration,
+  connectionTimeout: Duration)
+    extends State
+    with JavaScalaFutureSupport
+    with Logging {
+
+  private val eventualCuratorClient: Future[CuratorFramework] = createZookeeperClient()
+
+  /**
+    * Block until we have a connection to Zookeeper. If the max time is reached
+    * an exception will be thrown.
+    * *
+    * @param duration the max amount of time to wait
+    */
+  def await(duration: Duration): Unit = {
```

Comment:
The method name isn't totally clear.  How about something along the lines of `awaitZookeeperConnection`?